File: crates/scon-core/src/analysis.rs
use std::fs;
use std::path::{Path, PathBuf};

use crate::ast::{AstValue, LocalMember, ObjectBody};
use crate::error::{Error, ErrorCode, Result};
use crate::limits::LoadOptions;
use crate::source::{self, LineIndex, Token};
pub use crate::source::{Comment, CommentKind, SourcePosition, SourceRange, Utf16Position};
use crate::value::Value;

#[derive(Clone, Debug, Default)]
pub struct ParseOptions {
    pub file: Option<PathBuf>,
}

#[derive(Clone, Debug)]
pub struct FormatOptions {
    pub indent: usize,
}

impl Default for FormatOptions {
    fn default() -> Self {
        Self { indent: 2 }
    }
}

#[derive(Clone, Debug)]
pub struct Symbol {
    pub path: Vec<String>,
    pub range: SourceRange,
}

#[derive(Clone, Debug)]
pub struct Diagnostic {
    pub code: ErrorCode,
    pub message: String,
    pub file: Option<PathBuf>,
    pub range: Option<SourceRange>,
}

#[derive(Clone, Debug)]
pub struct ParsedDocument {
    pub file: Option<PathBuf>,
    pub line_index: LineIndex,
    pub tokens: Vec<Token>,
    pub comments: Vec<Comment>,
    pub symbols: Vec<Symbol>,
}

#[derive(Clone, Debug)]
pub struct Analysis {
    pub diagnostics: Vec<Diagnostic>,
    pub comments: Vec<Comment>,
    pub symbols: Vec<Symbol>,
    pub value: Option<Value>,
}

pub fn parse_source(source: &str, options: ParseOptions) -> Result<ParsedDocument> {
    let document = crate::parser::parse_str(source, options.file.clone())?;
    let line_index = LineIndex::new(source);
    let tokens = source::collect_tokens(source);
    let comments = source::comments_from_tokens(source, &line_index, &tokens);
    let symbols = collect_symbols(&document.body, source, &line_index, &[]);
    Ok(ParsedDocument {
        file: options.file,
        line_index,
        tokens,
        comments,
        symbols,
    })
}

pub fn analyze_source(source: &str, options: ParseOptions) -> Analysis {
    match parse_source(source, options) {
        Ok(parsed) => Analysis {
            diagnostics: Vec::new(),
            comments: parsed.comments,
            symbols: parsed.symbols,
            value: None,
        },
        Err(err) => Analysis {
            diagnostics: vec![diagnostic_from_error(&err, source)],
            comments: {
                let line_index = LineIndex::new(source);
                let tokens = source::collect_tokens(source);
                source::comments_from_tokens(source, &line_index, &tokens)
            },
            symbols: Vec::new(),
            value: None,
        },
    }
}

pub fn analyze_file(path: impl AsRef<Path>, options: LoadOptions) -> Analysis {
    let path = path.as_ref();
    let source = fs::read_to_string(path).unwrap_or_default();
    let parsed = parse_source(
        &source,
        ParseOptions {
            file: Some(path.to_path_buf()),
        },
    );
    let (comments, symbols) = match parsed {
        Ok(parsed) => (parsed.comments, parsed.symbols),
        Err(_) => {
            let line_index = LineIndex::new(&source);
            let tokens = source::collect_tokens(&source);
            (
                source::comments_from_tokens(&source, &line_index, &tokens),
                Vec::new(),
            )
        }
    };
    match crate::parse_file_with_options(path, options) {
        Ok(value) => Analysis {
            diagnostics: Vec::new(),
            comments,
            symbols,
            value: Some(value),
        },
        Err(err) => Analysis {
            diagnostics: vec![diagnostic_from_error(&err, &source)],
            comments,
            symbols,
            value: None,
        },
    }
}

pub fn format_source(source: &str, options: FormatOptions) -> Result<String> {
    crate::parser::parse_str(source, None)?;
    Ok(format_source_unchecked(source, options))
}

pub fn resolve_source(source: &str, options: ParseOptions) -> Result<Value> {
    let doc = crate::parser::parse_str(source, options.file)?;
    crate::eval::eval_document(doc, &mut crate::loader::NoopLoader)
}

pub fn resolve_file(path: impl AsRef<Path>, options: LoadOptions) -> Result<Value> {
    crate::parse_file_with_options(path, options)
}

pub fn get_path<'a>(value: &'a Value, path: &str) -> Result<&'a Value> {
    let path = parse_path_query(path)?;
    let mut current = value;
    for segment in &path {
        let Value::Object(object) = current else {
            return Err(Error::new(
                ErrorCode::TypeMismatch,
                "path segment requires object",
            ));
        };
        current = object.get(segment).ok_or_else(|| {
            Error::new(
                ErrorCode::MissingReference,
                format!("path {:?} is not defined", path),
            )
            .with_path(&path)
        })?;
    }
    Ok(current)
}

pub fn diagnostic_from_error(error: &Error, source: &str) -> Diagnostic {
    let line_index = LineIndex::new(source);
    let byte = line_index
        .byte_for_line_character(
            source,
            error.line.saturating_sub(1),
            error.column.saturating_sub(1),
        )
        .unwrap_or(0);
    let start = SourcePosition {
        line: error.line.saturating_sub(1),
        character: error.column.saturating_sub(1),
        byte,
    };
    let end = SourcePosition {
        line: start.line,
        character: start.character + 1,
        byte: start.byte.saturating_add(1),
    };
    Diagnostic {
        code: error.code,
        message: error.message.clone(),
        file: error.file.clone(),
        range: Some(SourceRange { start, end }),
    }
}

fn format_source_unchecked(source: &str, options: FormatOptions) -> String {
    let mut out = String::new();
    let mut indent = 0usize;
    for raw_line in source.lines() {
        let trimmed = raw_line.trim();
        if trimmed.is_empty() {
            out.push('\n');
            continue;
        }
        if starts_with_closer(trimmed) {
            indent = indent.saturating_sub(options.indent);
        }
        write_spaces(&mut out, indent);
        out.push_str(&format_line(trimmed));
        out.push('\n');
        if opens_block(trimmed) {
            indent += options.indent;
        }
    }
    if !out.ends_with('\n') {
        out.push('\n');
    }
    out
}

fn format_line(line: &str) -> String {
    let Some(eq) = find_unquoted(line, b'=') else {
        return line.to_string();
    };
    let left = line[..eq].trim_end();
    let right = line[eq + 1..].trim_start();
    format!("{left} = {right}")
}

fn starts_with_closer(line: &str) -> bool {
    matches!(line.as_bytes().first(), Some(b'}' | b']'))
}

fn opens_block(line: &str) -> bool {
    let code = strip_inline_comment(line).trim_end();
    code.ends_with('{') || code.ends_with('[')
}

fn strip_inline_comment(line: &str) -> &str {
    let mut in_string = false;
    let mut escaped = false;
    let bytes = line.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        let byte = bytes[i];
        if escaped {
            escaped = false;
            i += 1;
            continue;
        }
        match byte {
            b'\\' if in_string => escaped = true,
            b'"' => in_string = !in_string,
            b'#' if !in_string => return &line[..i],
            b'/' if !in_string && bytes.get(i + 1) == Some(&b'/') => return &line[..i],
            _ => {}
        }
        i += 1;
    }
    line
}

fn find_unquoted(line: &str, target: u8) -> Option<usize> {
    let mut in_string = false;
    let mut escaped = false;
    for (index, byte) in line.bytes().enumerate() {
        if escaped {
            escaped = false;
            continue;
        }
        match byte {
            b'\\' if in_string => escaped = true,
            b'"' => in_string = !in_string,
            byte if byte == target && !in_string => return Some(index),
            b'#' if !in_string => return None,
            b'/' if !in_string && line.as_bytes().get(index + 1) == Some(&b'/') => return None,
            _ => {}
        }
    }
    None
}

fn collect_symbols(
    body: &ObjectBody,
    source: &str,
    line_index: &LineIndex,
    parent_path: &[String],
) -> Vec<Symbol> {
    let mut symbols = Vec::new();
    for member in &body.members {
        let LocalMember::Field(field) = member else {
            continue;
        };
        let mut path = parent_path.to_vec();
        path.extend(field.path.iter().cloned());
        symbols.push(Symbol {
            path: path.clone(),
            range: line_index.source_range(source, field.path_span),
        });
        if let AstValue::Object { body, .. } = &field.value {
            symbols.extend(collect_symbols(body, source, line_index, &path));
        }
    }
    symbols
}

fn parse_path_query(path: &str) -> Result<Vec<String>> {
    let mut parts = Vec::new();
    let mut current = String::new();
    let mut chars = path.chars().peekable();
    while let Some(ch) = chars.next() {
        match ch {
            '.' if !current.is_empty() => {
                parts.push(std::mem::take(&mut current));
            }
            '"' if current.is_empty() => {
                while let Some(ch) = chars.next() {
                    match ch {
                        '"' => break,
                        '\\' => {
                            let Some(escaped) = chars.next() else {
                                return Err(Error::new(
                                    ErrorCode::InvalidEscape,
                                    "unterminated quoted path escape",
                                ));
                            };
                            current.push(escaped);
                        }
                        ch => current.push(ch),
                    }
                }
            }
            ch => current.push(ch),
        }
    }
    if !current.is_empty() {
        parts.push(current);
    }
    if parts.is_empty() {
        return Err(Error::new(ErrorCode::UnexpectedToken, "empty path"));
    }
    Ok(parts)
}

fn write_spaces(out: &mut String, count: usize) {
    for _ in 0..count {
        out.push(' ');
    }
}

File: typescript/packages/scon-core/src/analysis.ts
import type { AstMember, AstObject, AstPath, AstValue, Document } from "./ast.js";
import type { ErrorCode, Span } from "./error.js";
import { SconError } from "./error.js";
import { lex, type Token } from "./lexer.js";
import { parseDocument } from "./parser.js";
import { parseString } from "./resolver.js";
import type { SconValue } from "./value.js";

export type DiagnosticSeverity = "error" | "warning" | "information" | "hint";
export type SourcePosition = { line: number; column: number };
export type SourceRange = { start: SourcePosition; end: SourcePosition; span: Span };
export type Comment = { text: string; span: Span; range: SourceRange };
export type Diagnostic = {
  code: ErrorCode;
  message: string;
  severity: DiagnosticSeverity;
  file?: string;
  range?: SourceRange;
};
export type Symbol = { path: string[]; file?: string; range: SourceRange };
export type Definition = { path: string[]; file?: string; range: SourceRange };
export type ReferenceKind = "substitution" | "interpolation" | "objectSpread" | "arraySpread";
export type Reference = { path: string[]; kind: ReferenceKind; file?: string; range: SourceRange; target?: Definition };
export type IncludeReference = { path: string; file?: string; range: SourceRange; resolvedPath?: string };
export type ParsedSource = {
  file?: string;
  tokens: Array<Token & { range: SourceRange }>;
  comments: Comment[];
  symbols: Symbol[];
};
export type Analysis = {
  file?: string;
  parsed?: ParsedSource;
  diagnostics: Diagnostic[];
  comments: Comment[];
  symbols: Symbol[];
  definitions: Definition[];
  references: Reference[];
  includes: IncludeReference[];
  value?: SconValue;
};

export function parseSource(source: string, file?: string): ParsedSource {
  const document = parseDocument(source, file);
  const lineIndex = new LineIndex(source);
  const tokens = lex(source).map((token) => ({ ...token, range: lineIndex.range(token.span) }));
  return {
    file,
    tokens,
    comments: comments(tokens),
    symbols: collectSymbols(document.root, lineIndex, file, []),
  };
}

export function analyzeSource(source: string, file?: string): Analysis {
  const lineIndex = new LineIndex(source);
  let tokens: Array<Token & { range: SourceRange }> = [];
  try {
    tokens = lex(source).map((token) => ({ ...token, range: lineIndex.range(token.span) }));
    const document = parseDocument(source, file);
    const parsed = {
      file,
      tokens,
      comments: comments(tokens),
      symbols: collectSymbols(document.root, lineIndex, file, []),
    };
    const definitions = collectDefinitions(document.root, lineIndex, file, []);
    const references = collectReferences(document, lineIndex, file);
    resolveTargets(references, definitions);
    let value: SconValue | undefined;
    const diagnostics: Diagnostic[] = [];
    try {
      value = parseString(source);
    } catch (err) {
      diagnostics.push(diagnosticFromError(err, lineIndex, file));
    }
    return {
      file,
      parsed,
      diagnostics,
      comments: parsed.comments,
      symbols: parsed.symbols,
      definitions,
      references,
      includes: collectIncludes(document.root, lineIndex, file),
      value,
    };
  } catch (err) {
    return {
      file,
      diagnostics: [diagnosticFromError(err, lineIndex, file)],
      comments: comments(tokens),
      symbols: [],
      definitions: [],
      references: [],
      includes: [],
    };
  }
}

function collectSymbols(object: AstObject, lineIndex: LineIndex, file: string | undefined, prefix: string[]): Symbol[] {
  return object.members.flatMap((member) => {
    if (member.type !== "field") return [];
    const path = [...prefix, ...pathNames(member.path)];
    const nested = member.value.type === "object" ? collectSymbols(member.value.object, lineIndex, file, path) : [];
    return [{ path, file, range: lineIndex.range(member.path.span) }, ...nested];
  });
}

function collectDefinitions(object: AstObject, lineIndex: LineIndex, file: string | undefined, prefix: string[]): Definition[] {
  return object.members.flatMap((member) => {
    if (member.type !== "field") return [];
    const path = [...prefix, ...pathNames(member.path)];
    const nested = member.value.type === "object" ? collectDefinitions(member.value.object, lineIndex, file, path) : [];
    return [{ path, file, range: lineIndex.range(member.path.span) }, ...nested];
  });
}

function collectReferences(document: Document, lineIndex: LineIndex, file?: string): Reference[] {
  return collectObjectReferences(document.root, lineIndex, file);
}

function collectObjectReferences(object: AstObject, lineIndex: LineIndex, file?: string): Reference[] {
  return object.members.flatMap((member) => {
    switch (member.type) {
      case "objectSpread":
        return [reference(member.sub.path, "objectSpread", lineIndex, file)];
      case "field":
        return collectValueReferences(member.value, lineIndex, file);
      case "include":
        return [];
    }
  });
}

function collectValueReferences(value: AstValue, lineIndex: LineIndex, file?: string): Reference[] {
  switch (value.type) {
    case "substitution":
      return [reference(value.path, "substitution", lineIndex, file)];
    case "string":
      return value.parts.flatMap((part) => part.type === "interpolation" ? [reference(part.path, "interpolation", lineIndex, file)] : []);
    case "array":
      return value.items.flatMap((item) => item.type === "spread"
        ? [reference(item.sub.path, "arraySpread", lineIndex, file)]
        : collectValueReferences(item.value, lineIndex, file));
    case "object":
      return collectObjectReferences(value.object, lineIndex, file);
    default:
      return [];
  }
}

function collectIncludes(object: AstObject, lineIndex: LineIndex, file?: string): IncludeReference[] {
  return object.members.flatMap((member: AstMember) => {
    if (member.type === "include") return [{ path: member.path.value, file, range: lineIndex.range(member.span) }];
    if (member.type === "field" && member.value.type === "object") return collectIncludes(member.value.object, lineIndex, file);
    return [];
  });
}

function reference(path: AstPath, kind: ReferenceKind, lineIndex: LineIndex, file?: string): Reference {
  return { path: pathNames(path), kind, file, range: lineIndex.range(path.span) };
}

function resolveTargets(references: Reference[], definitions: Definition[]): void {
  const byPath = new Map(definitions.map((definition) => [definition.path.join("\0"), definition]));
  for (const reference of references) {
    reference.target = byPath.get(reference.path.join("\0"));
  }
}

function comments(tokens: Array<Token & { range: SourceRange }>): Comment[] {
  return tokens.filter((token) => token.kind === "comment").map((token) => ({
    text: token.text,
    span: token.span,
    range: token.range,
  }));
}

function diagnosticFromError(err: unknown, lineIndex: LineIndex, file?: string): Diagnostic {
  if (err instanceof SconError) {
    return {
      code: err.code,
      message: err.message,
      severity: "error",
      file,
      range: err.span ? lineIndex.range(err.span) : undefined,
    };
  }
  return { code: "Serde", message: err instanceof Error ? err.message : String(err), severity: "error", file };
}

function pathNames(path: AstPath): string[] {
  return path.segments.map((segment) => segment.value);
}

class LineIndex {
  private readonly lines = [0];

  constructor(private readonly source: string) {
    for (let index = 0; index < source.length; index++) {
      if (source[index] === "\n") this.lines.push(index + 1);
    }
  }

  range(span: Span): SourceRange {
    return { start: this.position(span.start), end: this.position(span.end), span };
  }

  private position(offset: number): SourcePosition {
    let line = 0;
    while (line + 1 < this.lines.length && this.lines[line + 1]! <= offset) line++;
    return { line, column: offset - this.lines[line]! };
  }
}
